terraform/modules/monitoring/monitoring.tf

```diff
@@ -0,0 +1,181 @@
+  metadata {
+    name = "monitor"
+  }
+}
+
+resource "kubernetes_service" "prometheus" {
+  metadata {
+    name      = "prometheus"
+    namespace = "monitor"
+  }
+  spec {
+    selector = {
+      app = kubernetes_pod.prometheus.metadata.0.labels.app
+    }
+    port {
+      port        = 9090
+      target_port = 9090
+    }
+
+    type = "LoadBalancer"
+  }
+}
+
+resource "kubernetes_pod" "prometheus" {
+  metadata {
+    name      = "prometheus"
+    namespace = "monitor"
+    labels = {
+      app = "Prometheus"
+    }
+  }
+
+  spec {
+    service_account_name = kubernetes_service_account.prometheus_account.metadata[0].name
+    container {
+      image = "prom/prometheus:latest"
+      name  = "prometheus"
+      args = [
+        "--config.file", "/config/prometheus.yml",
+        "--storage.tsdb.path", "/data"
+      ]
+      volume_mount {
+        name       = "config-volume"
+        mount_path = "/config"
+      }
+      volume_mount {
+        name       = "storage-volume"
+        mount_path = "/data"
+      }
+    }
+    security_context {
+      fs_group        = 2000
+      run_as_user     = 1000
+      run_as_non_root = true
+    }
+    volume {
+      name = "config-volume"
+      config_map {
+        name = kubernetes_config_map.prometheus_config.metadata[0].name
+      }
+    }
+    volume {
+      name = "storage-volume"
+
+      persistent_volume_claim {
+        claim_name = kubernetes_persistent_volume_claim.prometheus_disk_claim.metadata[0].name
+      }
+    }
+  }
+}
+
+resource "kubernetes_config_map" "prometheus_config" {
+  metadata {
+    name      = "prometheus-config"
+    namespace = "monitor"
+  }
+
+  data = {
+    "prometheus.yml" = file("${path.module}/prometheus.yml")
+  }
+}
+
+resource "kubernetes_cluster_role" "prometheus_cluster_role" {
+  metadata {
+    name = "prometheus-role"
+  }
+
+  rule {
+    // API group "" means the core API group.
+    api_groups = [""]
+    resources  = ["nodes", "services", "endpoints", "pods"]
+    verbs      = ["get", "list", "watch"]
+  }
+  rule {
+    api_groups = [""]
+    resources  = ["configmaps"]
+    verbs      = ["get"]
+  }
+}
+
+resource "kubernetes_service_account" "prometheus_account" {
+  metadata {
+    name      = "prometheus-account"
+    namespace = "monitor"
+  }
+}
+
+resource "kubernetes_cluster_role_binding" "prometheus_binding" {
+  metadata {
+    name = "prometheus_binding"
+  }
+
+  role_ref {
+    kind      = "ClusterRole"
+    name      = "prometheus_cluster_role"
+    api_group = "rbac.authorization.k8s.io"
+  }
+
+  subject {
+    kind      = "ServiceAccount"
+    name      = "prometheus-account"
+    namespace = "monitor"
+  }
+}
+
+resource "kubernetes_persistent_volume_claim" "prometheus_disk_claim" {
+  metadata {
+    name      = "prometheus-disk-claim"
+    namespace = "monitor"
+  }
+  spec {
+    access_modes = ["ReadWriteOnce"]
+    resources {
+      requests = {
+        storage = "5Gi"
```

Comment:
This is the volume that Prometheus is going to store time series data in, right? Any thoughts on whether 5 GB or 20 GB will be sufficient? I know that's a tricky capacity planning question given that we don't even yet know what kind of data we want to store here, but maybe we should make these parameters configurable so it's easy to grow these volumes in the future?

Reply:
These are total shots in the dark. My plan is to deploy and measure how much we're using. As far as configurability - I think it's pretty easy to edit the hardcoded value here in the future, or turn it into a var.

One potential stumbling block: The "claim" is actually at 5Gi rather than the full 20Gi of its persistent disk. During testing I set it to be much lower, worried that the claim wasn't becoming bound because for some reason it was asking for too much. Later I tried increasing the claim, and ran into some errors that it couldn't be resized. Seems like we may need to do some extra stuff to allow resizes.